b2luigi/batch/processes/gbasf2.py

```diff
@@ -0,0 +1,764 @@
+
+        # estimated number or processed events per second
+        evtpersec = get_setting("gbasf2_evtpersec", default=False, task=self.task)
+        if evtpersec is not False:
+            gbasf2_command_str += f" --evtpersec {evtpersec} "
+
+        # gbasf2 job priority
+        priority = get_setting("gbasf2_priority", default=False, task=self.task)
+        if priority is not False:
+            assert 0 <= priority <= 10, "Priority should be integer between 0 and 10."
+            gbasf2_command_str += f" --priority {priority} "
+
+        # gbasf2 job type (e.g. User, Production, ...)
+        jobtype = get_setting("gbasf2_jobtype", default=False, task=self.task)
+        if jobtype is not False:
+            gbasf2_command_str += f" --jobtype {jobtype} "
+
+        # additional basf2 options to use on grid
+        basf2opt = get_setting("gbasf2_basf2opt", default=False, task=self.task)
+        if basf2opt is not False:
+            gbasf2_command_str += f" --basf2opt='{basf2opt}' "
+
+        # optional string of additional parameters to append to gbasf2 command
+        gbasf2_additional_params = get_setting("gbasf2_additional_params", default=False, task=self.task)
+        if basf2opt is not False:
+            gbasf2_command_str += f" {gbasf2_additional_params} "
+
+        gbasf2_command = shlex.split(gbasf2_command_str)
+        return gbasf2_command
+
+    def _write_path_to_file(self):
+        """
+        Serialize and save the ``basf2.Path`` returned by ``self.task.create_path()`` to a python pickle file.
+        """
+        try:
+            path = self.task.create_path()
+        except AttributeError as err:
+            raise Exception(
+                "Gbasf2 batch process can only used with tasks that generate basf2 paths with "
+                "a ``create_path()`` method, e.g. are an instance of ``Basf2PathTask``."
+            ) from err
+
+        path.add_module("Progress")
+        b2pp.write_path_to_file(path, self.pickle_file_path)
+
+    def _create_wrapper_steering_file(self):
+        """
+        Create a steering file to send to the grid that executes the pickled
+        basf2 path from ``self.task.create_path()``.
+        """
+        # read a jinja2 template for the steerinfile that should execute the pickled path
+        template_file_path = os.path.join(self._file_dir, "templates/gbasf2_steering_file_wrapper.jinja2")
+        with open(template_file_path, "r") as template_file:
+            template = Template(template_file.read())
+            # replace some variable values in the template
+            steering_file_stream = template.stream(
+                pickle_file_path=self.pickle_file_path,
+                max_event=get_setting("max_event", default=0, task=self.task),
+            )
+            # write the template with the replacements to a new file which should be sent to the grid
+            steering_file_stream.dump(self.wrapper_file_path)
+
+    def _download_dataset(self):
```

Comment:
Maybe I do not get it, but why do you need an additional `gbasf2_output_directory`? Is the `result_dir` not enough - or put other way: why are you not using the locations of the file paths in the targets of `output()`?
Sorry, do not know gbasf2 enough,

Reply:
Thanks for spotting that, that's something where I was interested in your opinion but forgot to comment on that explicitly. In most cases that I had seen so far, the `output()` of a task doesn't influence the what happens in the `run()`, it is only used to define the targets of the task, but the target paths are acutally set e.g. in your `create_path()` in the `RootOutput` module or whatever. So I felt it's not supposed to be used for *setting* the output location. But the `create_path()` has also no influence where the downloaded gbasf2 dataset will finally go, so I decided to set it via a setting. It is set by default to the path you get from `get_output_file_name(...)`, but there might be users who don't want to use the output helpers and `result_dir` and for them I decided to provide this setting.

I feared ising `output()` to set the output directory path might also take away some flexibility from the user in using some other targets in the `output()` which are not the dataset directory. E.g. in a commit yesterday I added a `Gbasf2GridProjectTarget` which checks if the dataset exists on the grid and is finished, for jobs where the dataset download is disabled via a setting. But well, in that case particular case I would never execute the download code, so the fact the `output` is not a file system target wouldn't be a problem there, now that I think about it.

The output handling is something that I wasn't sure so I am happy about your suggestions. I thought maybe I could also add some helper functions for the user. But if you think my fears are no reason to worry about, I might just use `output()`.

gbasf2 is not complicated in that, thare is a `gb2_ds_get` command which takes a project name and by defaults downloads  the root files into `./<project name>/<output root files>`. You can also use a command line option (`-i`) to download it into another directory, but it will still have the project name in a subdirectory. I don't use that CLI option but just set the `cwd` argument in subprocess and then move/rename the directory to its final location.

Reply:
I guess I do not understand all of this correctly (not due to the fact you explaining it wrongly, but due to the fact that I never touched any of that), but I have the feeling you are somehow "mis-using" luigi here.
The targets you define in your Task (basically what you return in `output()`) are there for two reasons in luigi:
1. define for any dependent task where to find its input files
2. tell luigi when a task is finished, basically iff all targets return true on their `exist()` call.

If I understand your comment correctly, the following could happend:
* a user defines some `LocalTarget` with a specific path in the `output()` call
* in the created path, the user specifies another output file name (which is already not how it should work)
* your code will download the created file into the folder specified by `gbasf2_output_directory`, or by default to `get_output_file_name(get_setting("gbasf2_project_name_prefix", task=self.task))`

If you now specify a `result_dir` of anything not being `gbasf2_project_name_prefix` (or `gbasf2_output_directory`), this task will never be finished for luigi, because its targets never exist.

Questions from my side:
* how does gbasf2 determine what it should download, e.g. what is in the "project dataset"?
* you said (in some command above) the full absolute file paths are regenerated inside of the grid environment. Could one just write files in the grid job already to their final location (as given by the path in the LocalTarget) and then download them and copy them exactly to where they belong?
* I did not quite understand your sentence "In most cases that I had seen so far, the output() of a task doesn't influence the what happens in the run()". Because looking into our examples (https://b2luigi.readthedocs.io/en/stable/advanced/basf2-examples.html) we tell the user to use the function "self.get_output_file_name" to specify the output file name.

If using "get_output_file_name" is a problem on grid jobs (due to whatever reason), I would argue to only allow for custom targets, such as your already defined target without download and to add one with download. It would be clearer to the user that these are not typical files but more a "flag" that the job is done. Still, you could give them a file `path` of the downloaded location (maybe controlled by your parameter) that would make them usable in any task depending this one.

Example:

```python
class MyTask(...):
    def output(self):
        yield Gbasf2GridProjectTargetWithDownload("ntuples.root") # this could check only locally 

    def create_path(self):
        ... RootOutput("ntuples.root")....

```

Still, I would definitely vote (if possible) for a solution like:

```python
class MyTask(...):
    def output(self):
        yield self.add_to_output("ntuples.root")
        # or, if the user really wants
        yield LocalTarget(.... ntuples.root ....)

    def create_path(self):
        ... RootOutput(self.get_output_file_name("ntuples.root"))....
   
   # or if this does not work

    def create_path(self):
        ... RootOutput("ntuples.root")....

```

The reason I strongly vote for this, is that this is what luigi is all about in its core: define targets and the definition of done for a task is, when someone has written exactly this target. Its much cleaner then.

So the way I would handle it: 
on downloading, loop through all targets in `task.output()`, check if they actually have a path, download the dataset into a temporary folder and move the files around.
You said above, that gbasf2 will create not a single output, but actually multiple ones (e.g. called `ntuple_X.root`). You could work around that by assuming the file paths given in the "output" are actually folder names. So in the example code above I would expect to have a folder called "ntuples.root" in my result_dir (with the usual parameters before that), which includes all the ntuples files.

What do you think about that? Does this fit your idea and how gbasf2 works? Sorry if I give stupid comments, please correct me.

Reply:
Thanks for the detailed answer and suggestions!

1. **how does gbasf2 determine what it should download, e.g. what is in the "project dataset"?**

    On the grid, files can be accessed via a logical file name (LFN) which is a unix-like path. The output data of all user jobs is actually stored unter  `/belle/user/<user name>/<project>/file`. I just found out on confluence that there's also the possiblity to have subdirectories in special cases, I just copied a section:

    > `/belle/user/<user name>/<project>/file`
    > or
    > `/belle/user/<user name>/<project>/<subdir>/file`
    >
    > Subdirectories can be either
    > - the ones created by the option '--subdir_nfiles'
    > - the ones created following the DataDescription metadata in multi-output jobs
    >
    > File names (the basename) are combination of
    > - user-defined output file name
    > - job number

    The gbasf2 submit command also has cli option to store the outputs in a entirely different path, e.g when you want to provide data to your subgroup or do data production. But I provide only the project name in my submission command and expect the outputs to be under the user directory.

    When you then do the dataset download command, one way is to use the full LFN path to a file or directory containing files. Another way is to just provide the project name to that command and it will look for a directory with that name under `/belle/users/<user>`. With the `-u` option you can choose another user than yourself, which is the default. I set that explicitly in most commands though not really needed. Exactly how gbasf2 to figures things out internally I don't know, I just know the imperfectly documented user side of things. Here's the help string of the download command: https://pastebin.com/0P9xL0P6

2. **path regeneration**
    The path regeneration happens in the "job input sandbox". Then there's also the "job output sandbox" which contains log files from the helper functions that gbasf2 uses to execute the steering file on the grid and has a maximum size of around 10mb. But the paths are not regenerated on the grid where the datasets are stored. I still did a simple test where I added two subdirectories locally and send a job to the grid with `filename="./testdir/testsubdir/ntuple.root"` and it failed because basf2 couldn't find that directory.

3. **What I meant**
    I am not sure where exactly our misunderstanding lies. By using something like `output _path = self.get_output_file_name(basename)` the user in the examples explicitly set the output path in the `create_path()` and the path returned by `get_output_file_name()` does not depend on what is returned by `output`. Though by using `add_to_output()`, we make sure the target matches the correct output path. I had doubts about omitting the `get_output_file_name()`  part, but I thought that over and think that should be no problem.

I don't see any reason what your ideas that you mentioned under "*So the way I would handle it*" wouldn't work, the simple idea of the directories named like the root files is what I was missing. I'll think through your examples carefully on a fresh mind.

Reply:
In case of multiple outputs, which will each be seperately enumerated in the output dataset, I can use wildcard in the `gb2_ds_get` command (see pastebin above) to download each subset seperately, e.g. `gb2_ds_get <project>/ntupleA_*.root` and `gb2_ds_get <project>/ntupleB_*.root` into the dirs `get_output_file_name("ntuple[A|B].root")`. Or I could do my own globbing, both should be not to hard. I like your ideas.